module/zfs/arc.c

```diff
@@ -6058,7 +6058,9 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 				    spa, NULL, zb, NULL, 0, 0);
 			}
 			if (rc != 0) {
-				arc_buf_destroy(buf, private);
+				(void) remove_reference(hdr, hash_lock,
+				    private);
+				arc_buf_destroy_impl(buf);
```

Comment:
The old code (pre-encryption) simply VERIFY'ed that the arc_buf_alloc_impl() succeeds.  While this is primitive, it "works".  We definitely need to exercise this new error path to ensure that it works correctly.

There's a similar decompression error that can happen in arc_read_done(), which is addressed by https://github.com/openzfs/openzfs/pull/601.  That also has test code (of a style not accepted in ZoL) to exercise this the error path.

Reply:
I just spoke to Brian about this and I'm going to look to add zinject support for encryption failures.

Reply:
Adding coverage to ztest for encryption failures wouldn't be a bad idea either to improve our testing off these error paths.